src/RendezvousTracker.ts

```diff
@@ -0,0 +1,268 @@
+        this.emitter.on(eventName, handler);
+        return () => {
+            if (this.emitter !== undefined) {
+                this.emitter.removeListener(eventName, handler);
+            }
+        };
+    }
+
+    private emit(eventName: 'rendezvous-event', data?) {
+        this.emitter.emit(eventName, data);
+    }
+
+    /**
+     * Sets up the debug functions used to locate the client files and lines
+     */
+    public setDebuggerFileConversionFunctions(convertDebuggerLineToClientLine, convertDebuggerPathToClient) {
+        this.convertDebuggerLineToClientLine = convertDebuggerLineToClientLine;
+        this.convertDebuggerPathToClient = convertDebuggerPathToClient;
+    }
+
+    /**
+     * Used to set wether the rendezvous should be filtered from the console output
+     * @param outputLevel the consoleOutput from the launch config
+     */
+    public setConsoleOutput(outputLevel: string) {
+        this.filterOutLogs = !(outputLevel === 'full');
+    }
+
+    /**
+     * Takes the debug output from the device and parses it for any rendezvous information.
+     * Also if consoleOutput was not set to 'full' then any rendezvous output will be filtered from the output.
+     * @param logLine
+     * @returns The debug output after parsing
+     */
+    public processLogLine(logLine: string): string {
+        let dataChanged = false;
+        let lines = logLine.split('\n');
+        let normalOutput = '';
+
+        lines.map((line) => {
+            let match;
+            // see the following for an explanation for this regex: https://regex101.com/r/In0t7d/5
+            if (match = /\[sg\.node\.(BLOCK|UNBLOCK)\s{0,}\] Rendezvous\[(\d+)\](?:\s\w+\n|\s\w{2}\s(.*)\((\d+)\)|[\s\w]+(\d+\.\d+)+|\s\w+)/g.exec(line)) {
+                let [fullMatch, type, id, fileName, lineNumber, duration] = match;
+                if (type === 'BLOCK') {
+                    // detected the start of a rendezvous event
+                    this.rendezvousBlocks[id] = {
+                        fileName: this.updateClientPathMap(fileName, lineNumber),
+                        lineNumber: lineNumber
+                    };
+                } else if (type === 'UNBLOCK' && this.rendezvousBlocks[id]) {
+                    // detected the completion of a rendezvous event
+                    dataChanged = true;
+                    let blockInfo = this.rendezvousBlocks[id];
+                    let clientLineNumber: string = this.clientPathsMap[blockInfo.fileName].clientLines[blockInfo.lineNumber].toString();
+
+                    if (this.rendezvousHistory[blockInfo.fileName]) {
+                        // file is in history
+                        if (this.rendezvousHistory[blockInfo.fileName][clientLineNumber]) {
+                            // line is in history, just update it
+                            (this.rendezvousHistory[blockInfo.fileName][clientLineNumber] as RendezvousLineInfo).totalTime += this.getTime(duration);
```

Comment:
Please find a way to not need to type cast this object. I think we can make the interfaces more clear instead.

Reply:
Done